From 4352cc24dafb96add0c9f1530ecbe4b0ad77507b Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Mon, 8 Sep 2025 21:18:05 +0200
Subject: [PATCH 181/553] QWindowsVistaAnimation: adjust painting of the
 animated image

Drawing an image with a dpr != 1.0 might lead to rounding errors of the
sourceRect which results in jittering during the animation. Therefore
calculate the exact source rect and pass it to QPainter::drawImage().

Task-number: QTBUG-139307
Pick-to: 6.10.0 6.9 6.8
Change-Id: Ia61ae2b3645b4508b02c17a20bfbb5171ccb749b
Reviewed-by: Richard Moe Gustavsen <richard.gustavsen@qt.io>
(cherry picked from commit ed3cbc1bd7702deed67b0cb142ccde133adfbbd8)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/plugins/styles/modernwindows/qwindowsvistaanimation.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/plugins/styles/modernwindows/qwindowsvistaanimation.cpp b/src/plugins/styles/modernwindows/qwindowsvistaanimation.cpp
index 1f070ae323a..fcd47864043 100644
--- a/src/plugins/styles/modernwindows/qwindowsvistaanimation.cpp
+++ b/src/plugins/styles/modernwindows/qwindowsvistaanimation.cpp
@@ -14,7 +14,10 @@ bool QWindowsVistaAnimation::isUpdateNeeded() const
 
 void QWindowsVistaAnimation::paint(QPainter *painter, const QStyleOption *option)
 {
-    painter->drawImage(option->rect, currentImage());
+    const auto img = currentImage();
+    const auto sz = QSizeF(option->rect.size()) * img.devicePixelRatio();
+    QRectF sourceRect(0, 0, sz.width(), sz.height());
+    painter->drawImage(option->rect, img, sourceRect);
 }
 
 QT_END_NAMESPACE
-- 
2.51.2

